skyrl-train/examples/mini_swe_agent/mini_swe_generator.py

```diff
@@ -0,0 +1,238 @@
+import asyncio
+from typing import Dict, List, Optional, Any, Tuple
+from omegaconf import DictConfig
+import yaml
+import traceback
+import ray
+import json
+import threading
+from pathlib import Path
+
+from minisweagent.models import get_model
+from minisweagent.agents.default import DefaultAgent
+from minisweagent.run.utils.save import save_traj
+from minisweagent.config import get_config_path
+from .mini_swe_utils import evaluate_trajectory, get_sb_environment
+
+from skyrl_train.generators.skyrl_gym_generator import SkyRLGymGenerator, GeneratorOutput, GeneratorInput
+from skyrl_train.inference_engines.base import ConversationType
+from skyrl_train.inference_engines.inference_engine_client import InferenceEngineClient
+from skyrl_train.inference_engines.utils import get_sampling_params_for_backend
+
+_OUTPUT_FILE_LOCK = threading.Lock()
+
+
+def update_preds_file(output_path: Path, instance_id: str, model_name: str, result: str):
+    """Update the output JSON file with results from a single instance."""
+    with _OUTPUT_FILE_LOCK:
+        output_data = {}
+        if output_path.exists():
+            output_data = json.loads(output_path.read_text())
+        output_data[instance_id] = {
+            "model_name_or_path": model_name,
+            "instance_id": instance_id,
+            "model_patch": result,
+        }
+        output_path.write_text(json.dumps(output_data, indent=2))
+
+
+@ray.remote(num_cpus=0.01)
+def init_and_run(instance, litellm_model_name, sweagent_config, generator_cfg, data_source, sampling_params):
+    from loguru import logger
+
+    model_config = sweagent_config.get("model", {})
+    model_config.setdefault("model_kwargs", {}).update(sampling_params)
+    model = get_model(litellm_model_name, model_config)
+
+    agent = None
+    env = None
+    extra_info = None
+    result = None
+    reward = 0
+    error = None
+    try:
+        env = get_sb_environment(sweagent_config, instance, data_source)
+        agent = DefaultAgent(model, env, **sweagent_config.get("agent", {}))
+        exit_status, result = agent.run(instance["problem_statement"])  # type: ignore[arg-type]
+    except Exception as e:
+        logger.error(f"Error processing instance {instance['instance_id']}: {e}", exc_info=True)
+        exit_status, result = type(e).__name__, str(e)
+        error = str(e)
+        extra_info = {"traceback": traceback.format_exc()}
+    finally:
+        path = Path(generator_cfg.miniswe_traj_dir)
+        path.mkdir(parents=True, exist_ok=True)
+        path = path / (str(instance["instance_id"]) + ".json")
```

Comment:
It might be more idiomatic / readable to do `f"{instance["instance_id"]}.json"` here